pkg/noderesourcetopology/match.go

```diff
@@ -0,0 +1,358 @@
+			// if can't find requested resource on the node - skip (don't set it as available NUMA node)
+			// if unfound resource has 0 quantity probably this numa node can be considered
+			if !ok && quantity.Cmp(zeroQuantity) != 0 {
+				continue
+			}
+			// Check for the following:
+			// 1. set numa node as possible node if resource is memory or Hugepages (until memory manager will not be merged and
+			// memory will not be provided in CRD
+			// 2. set numa node as possible node if resource is cpu and it's not guaranteed QoS, since cpu will flow
+			// 3. set numa node as possible node if zero quantity for non existing resource was requested (TODO check topology manaager behaviour)
+			// 4. otherwise check amount of resources
+			if resource == v1.ResourceMemory ||
+				strings.HasPrefix(string(resource), string(v1.ResourceHugePagesPrefix)) ||
+				resource == v1.ResourceCPU && qos != v1.PodQOSGuaranteed ||
+				quantity.Cmp(zeroQuantity) == 0 ||
+				numaQuantity.Cmp(quantity) >= 0 {
+				resourceBitmask.Add(numaNode.NUMAID)
+			}
+		}
+		bitmask.And(resourceBitmask)
+	}
+}
+
+func (ph PodLevelResourceHandler) PolicyFilter(pod *v1.Pod, zones topologyv1alpha1.ZoneList) *framework.Status {
+	containers := []v1.Container{}
+	containers = append(pod.Spec.InitContainers, pod.Spec.Containers...)
+
+	resources := make(v1.ResourceList)
+
+	for _, container := range containers {
+		for resource, quantity := range container.Resources.Requests {
+			if quan, ok := resources[resource]; ok {
+				quantity.Add(quan)
+			}
+			resources[resource] = quantity
+		}
+	}
+
+	nodes := createNUMANodeList(zones)
+	bitmask := bm.NewEmptyBitMask()
+	bitmask.Fill()
+	checkResourcesForNUMANodes(bitmask, nodes, resources, v1qos.GetPodQOS(pod))
+
+	if bitmask.IsEmpty() {
+		// definitly we can't align container, so we can't align a pod
+		return framework.NewStatus(framework.Unschedulable, fmt.Sprintf("Can't align pod: %s", pod.Name))
+	}
+	return nil
+
+}
+
+func createNUMANodeList(zones topologyv1alpha1.ZoneList) NUMANodeList {
+	nodes := make(NUMANodeList, 0)
+	for _, zone := range zones {
+		if zone.Type == "Node" {
+			var numaID int
+			fmt.Sscanf(zone.Name, "node-%d", &numaID)
+			resources := extractResources(zone)
+			nodes = append(nodes, NUMANode{NUMAID: numaID, Resources: resources})
+		}
+	}
+	return nodes
+}
+
+// Filter Now only single-numa-node supported
+func (tm *NodeResourceTopologyMatch) Filter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
+	if nodeInfo.Node() == nil {
+		return framework.NewStatus(framework.Error, fmt.Sprintf("Node is nil %s", nodeInfo.Node().Name))
+	}
+	if v1qos.GetPodQOS(pod) == v1.PodQOSBestEffort {
+		return nil
+	}
+
+	nodeName := nodeInfo.Node().Name
+
+	topologyPolicies := getTopologyPolicies(tm.nodeTopologies, nodeName)
+	for _, policyName := range topologyPolicies {
+		if handler, ok := tm.topologyPolicyHandlers[policyName]; ok {
```

Comment:
Given that the two objects are actually identical, and there is no struct fields associated with each, I'd suggest just create 2 stateless PolicyFilter() methods  (e.g., `PodLevelFilter()` and `SingleNUMANodeFilter()`)to handle them individually.

Reply:
Do you suggest to do it with map[string]interface{} or just compare policyName and call appropriate handler?

Reply:
We can define a function type:

```go
type PolicyFilter func(*v1.Pod, topologyv1alpha1.ZoneList) *framework.Status
```

And then register each function implementation in the map: `map[apiconfig.TopologyManagerPolicy]PolicyFilter`.